File: app/encv-mobile/src/types/appResult.ts
// AppResult 借鉴自 nuclear-boy AppResult.kt (sealed class AppResult<out T>)。
//
// TS 版用 discriminated union 实现：
//   - { ok: true; data: T }    → Success
//   - { ok: false; error: AppError } → Failure
//
// 配合类型守卫 isOk() / isFailure() 让消费方写链式 .OnSuccess / .OnFailure
// 时编译器能正确收窄类型。

import type { AppError } from "./appError";

/** AppResult<T> —— Success/Failure 联合类型。 */
export type AppResult<T> = { ok: true; data: T } | { ok: false; error: AppError };

/** 构造 Success 结果。 */
export function appOk<T>(data: T): AppResult<T> {
  return { ok: true, data };
}

/** 构造 Failure 结果。 */
export function appErr<T = never>(error: AppError): AppResult<T> {
  return { ok: false, error };
}

/** 类型守卫：是否为 Success。 */
export function isOk<T>(r: AppResult<T>): r is { ok: true; data: T } {
  return r.ok === true;
}

/** 类型守卫：是否为 Failure。 */
export function isFailure<T>(r: AppResult<T>): r is { ok: false; error: AppError } {
  return r.ok === false;
}

/** 取值或 fallback。 */
export function getOrElse<T>(r: AppResult<T>, fallback: T): T {
  return r.ok ? r.data : fallback;
}

/** Map 链式转换（Failure 透传，借鉴 nuclear-boy L24-27）。 */
export function mapResult<T, U>(r: AppResult<T>, transform: (data: T) => U): AppResult<U> {
  return r.ok ? appOk(transform(r.data)) : (r as AppResult<U>);
}

/** OnSuccess 副作用钩子（不改结果，借鉴 nuclear-boy L29-32）。 */
export function onOk<T>(r: AppResult<T>, action: (data: T) => void): AppResult<T> {
  if (r.ok) action(r.data);
  return r;
}

/** OnFailure 副作用钩子（不改结果，借鉴 nuclear-boy L34-37）。 */
export function onErr<T>(r: AppResult<T>, action: (error: AppError) => void): AppResult<T> {
  if (!r.ok) action(r.error);
  return r;
}

/** RunCatching —— 把可能 throw 的 async fn 包成 AppResult。
 *  借鉴 nuclear-boy AppResult.kt L44-53 runCatching。
 *
 *  与 Go RunCatching 行为对齐：
 *   - 返回成功 → { ok: true, data }
 *   - throw AppError → { ok: false, error }（直接复用）
 *   - throw 普通 Error → { ok: false, error: AppErrorUnknown + msg }
 *   - throw ToolError-like 对象 → 由调用方在 fn 内预先转为 AppError
 *
 *  @example
 *    const r = await runCatchingAsync(async () => await fetchXxx())
 *    if (isFailure(r)) showToast(r.error.humanMessage)
 */
export async function runCatchingAsync<T>(fn: () => Promise<T>): Promise<AppResult<T>> {
  try {
    const data = await fn();
    return appOk(data);
  } catch (e: unknown) {
    return appErr(toAppError(e));
  }
}

/** 同步版 runCatching。 */
export function runCatching<T>(fn: () => T): AppResult<T> {
  try {
    return appOk(fn());
  } catch (e: unknown) {
    return appErr(toAppError(e));
  }
}

/** 把任意 thrown 值转 AppError。
 *  - 已是 AppError → 直接返回
 *  - Error 有 name === 'AbortError' → UserCancelled
 *  - Error message 含 timeout/connection refused/401/402/429/HTTP 5xx → 启发式分类
 *  - 兜底 → Unknown
 */
function toAppError(e: unknown): AppError {
  if (isAppError(e)) return e;
  if (e instanceof Error) {
    const msg = e.message;
    const lower = msg.toLowerCase();
    if (e.name === "AbortError") return { type: "UserCancelled", message: msg, humanMessage: "🚫 已取消", isRetryable: false };
    if (lower.includes("timeout")) return { type: "NetworkTimeout", message: msg, humanMessage: "⏱️ 网络超时，请重试", isRetryable: true };
    if (lower.includes("connection refused") || lower.includes("no such host") || lower.includes("ssl"))
      return { type: "NetworkUnavailable", message: msg, humanMessage: "🌐 网络不可用", isRetryable: true };
    if (lower.includes("401")) return { type: "ApiKeyInvalid", message: msg, humanMessage: "🔑 API Key 无效", isRetryable: false };
    if (lower.includes("402")) return { type: "InsufficientBalance", message: msg, humanMessage: "💰 余额不足", isRetryable: false };
    if (lower.includes("429")) return { type: "RateLimited", message: msg, humanMessage: "🚦 请求过于频繁", isRetryable: true };
    if (msg.startsWith("HTTP 5") || lower.includes("internal server error"))
      return { type: "ServerError", message: msg, humanMessage: "🔧 服务端错误", isRetryable: true };
    return { type: "Unknown", message: msg, humanMessage: "❓ 未知错误", isRetryable: false };
  }
  return { type: "Unknown", message: String(e), humanMessage: "❓ 未知错误", isRetryable: false };
}

/** 类型守卫：判断任意值是否为 AppError。
 *  鸭子类型：必须有 type / message / humanMessage / isRetryable 四个字段。
 */
function isAppError(e: unknown): e is AppError {
  return typeof e === "object" && e !== null && "type" in e && "message" in e && "humanMessage" in e && "isRetryable" in e;
}

File: internal/tools/app_result.go
// Package tools 提供 AppResult<T> 借鉴自 nuclear-boy AppResult.kt (sealed class)。
//
// Go 习惯用 (value, error) tuple，但核 boy 的 AppResult 提供了 3 个独特的优势：
//   1. 显式 IsSuccess/IsFailure 判别（比 err != nil 更易读）
//   2. Map / OnSuccess / OnFailure 链式组合（避免嵌套 if-err）
//   3. 携带结构化 AppError（不只是字符串）
//
// 此文件提供 struct-based AppResult 作为可选 helper，调用方仍可继续用 (T, error) 风格。
package tools

// AppResult 借鉴自 nuclear-boy AppResult.kt L7-54。
//
// 判别模式：
//   - Success（T != 零值）→ IsSuccess=true
//   - Success 但 T 是零值 → 需配合 Error==nil 判定（典型 case：bool false）
//   - Failure → IsSuccess=false，Error 必有值
type AppResult[T any] struct {
	// Data 成功时的值；失败时为 T 类型的零值。
	Data T
	// Error 失败时的 AppError；成功时为 nil。
	Error *AppError
}

// NewAppResultSuccess 构造成功结果。
func NewAppResultSuccess[T any](data T) AppResult[T] {
	return AppResult[T]{Data: data}
}

// NewAppResultFailure 构造失败结果。
func NewAppResultFailure[T any](ae *AppError) AppResult[T] {
	var zero T
	return AppResult[T]{Data: zero, Error: ae}
}

// IsSuccess 是否成功。
func (r AppResult[T]) IsSuccess() bool { return r.Error == nil }

// IsFailure 是否失败。
func (r AppResult[T]) IsFailure() bool { return r.Error != nil }

// GetOrNull 取值或 nil。
func (r AppResult[T]) GetOrNull() T {
	if r.IsFailure() {
		var zero T
		return zero
	}
	return r.Data
}

// GetOrElse 取值或 fallback。
func (r AppResult[T]) GetOrElse(fallback T) T {
	if r.IsFailure() {
		return fallback
	}
	return r.Data
}

// Map 链式转换（Success 走 transform，Failure 直接透传）。
// 借鉴 nuclear-boy AppResult.kt L24-27。
func Map[T, U any](r AppResult[T], transform func(T) U) AppResult[U] {
	if r.IsFailure() {
		return NewAppResultFailure[U](r.Error)
	}
	return NewAppResultSuccess(transform(r.Data))
}

// OnSuccess 副作用钩子（不影响结果，借鉴 nuclear-boy L29-32）。
// 用于打日志/上报指标，不改 Error 状态。
func (r AppResult[T]) OnSuccess(action func(T)) AppResult[T] {
	if r.IsSuccess() {
		action(r.Data)
	}
	return r
}

// OnFailure 副作用钩子（不影响结果，借鉴 nuclear-boy L34-37）。
func (r AppResult[T]) OnFailure(action func(*AppError)) AppResult[T] {
	if r.IsFailure() {
		action(r.Error)
	}
	return r
}

// RunCatching 把可能 panic 的闭包包装为 AppResult。
// 借鉴 nuclear-boy AppResult.kt L44-53 runCatching。
//
// 与 Go 原生 (val, err) 的区别：
//   - 把抛出的 err 升级为 *AppError
//   - *AppError 直接复用；*ToolError 走 FromToolError 升级
//   - 其他 err 包装为 AppErrorUnknown，message = err.Error()
//
// 用法：
//
//	result := tools.RunCatching(func() (string, error) {
//	    return doSomething()
//	})
//	if result.IsFailure() { ... }
//
// 需要更精细的分类（network/timeout/401/5xx 等），调用方在 fn 内部预先调用
// agent.ClassifyException 把 err 转 AppError 再返回。
func RunCatching[T any](fn func() (T, error)) AppResult[T] {
	val, err := fn()
	if err == nil {
		return NewAppResultSuccess(val)
	}
	// *AppError 直接复用
	if ae, ok := err.(*AppError); ok {
		return NewAppResultFailure[T](ae)
	}
	// *ToolError 升级为 AppError
	if te, ok := err.(*ToolError); ok {
		return NewAppResultFailure[T](FromToolError(te))
	}
	// 其他 err 包装为 Unknown
	return NewAppResultFailure[T](NewAppError(AppErrUnknown, err.Error()))
}
